include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,803 @@
 
 #include "nbl/core/IReferenceCounted.h"
 #include "nbl/core/util/bitflag.h"
+#include "nbl/core/containers/CMemoryPool.h"
 
 #include "nbl/video/decl/IBackendObject.h"
 
 
 namespace nbl::video
 {
+class IGPUCommandBuffer;
 
 class NBL_API IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
 {
+public:
+    static inline constexpr uint32_t COMMAND_SEGMENT_SIZE = 128u << 10u;
+
+    enum E_CREATE_FLAGS : uint32_t
+    {
+        ECF_NONE = 0x00,
+        ECF_TRANSIENT_BIT = 0x01,
+        ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
+        ECF_PROTECTED_BIT = 0x04
+    };
+
+private:
+    static inline constexpr uint32_t COMMAND_ALIGNMENT = 64u;
+
+    static inline constexpr uint32_t COMMAND_SEGMENT_ALIGNMENT = 64u;
+
+    static inline constexpr uint32_t MAX_COMMAND_SEGMENT_BLOCK_COUNT = 16u;
+    static inline constexpr uint32_t COMMAND_SEGMENTS_PER_BLOCK = 256u;
+    static inline constexpr uint32_t MIN_POOL_ALLOC_SIZE = COMMAND_SEGMENT_SIZE;
+
+public:
+    class CCommandSegment;
+
+    class alignas(COMMAND_ALIGNMENT) ICommand
+    {
+    public:
+        virtual ~ICommand() {}
+
+        // static void* operator new(std::size_t size) = delete;
+        static void* operator new[](std::size_t size) = delete;
+        // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+        static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+        // static void operator delete  (void* ptr) = delete;
+        static void operator delete[](void* ptr) = delete;
+        static void operator delete  (void* ptr, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::align_val_t al) = delete;
+        static void operator delete  (void* ptr, std::size_t sz) = delete;
+        static void operator delete[](void* ptr, std::size_t sz) = delete;
+        static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+        inline uint32_t getSize() const { return m_size; }
+
+    protected:
+        ICommand(uint32_t size) : m_size(size) {}
+
+    private:
+        friend CCommandSegment;
+
+        const uint32_t m_size;
+    };
+
+    template <class CRTP>
+    class NBL_FORCE_EBO IFixedSizeCommand : public IGPUCommandPool::ICommand
+    {
     public:
-        enum E_CREATE_FLAGS : uint32_t
+        template <typename... Args>
+        static uint32_t calc_size(const Args&...)
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            return core::alignUp(sizeof(CRTP), alignof(CRTP));
```

Comment:
`sizeof(T)` is always already aligned to `alignof(T)`

Reply:
Oh yeah, `sizeof(T)` should always be a multiple of `alignof(T)`, at least, for a non-packed struct.
Removing..